base/task.jl

```diff
@@ -336,32 +336,42 @@ end
 
 ## lexically-scoped waiting for multiple items
 
+function schedule_result(t, rs)
+    schedule(Task(() -> begin
+        ex = nothing
+        try
+            wait(t)
+        catch e
+            ex = e
+        finally
+            put!(rs, ex)
+        end
+    end))
+end
+
 function sync_end(c::Channel{Any})
-    local c_ex
-    while isready(c)
-        r = take!(c)
-        if isa(r, Task)
-            _wait(r)
-            if istaskfailed(r)
-                if !@isdefined(c_ex)
-                    c_ex = CompositeException()
-                end
-                push!(c_ex, TaskFailedException(r))
-            end
-        else
-            try
-                wait(r)
-            catch e
-                if !@isdefined(c_ex)
-                    c_ex = CompositeException()
+    try
+        n = 0
+        isready(c) || return
+        while true
+            t = take!(c)
+            if t isa Exception                         # Exception from monitor. Collect
+                c_ex = CompositeException([t])         # any other exceptions and throw
+                while isready(c)
```

Comment:
I am not sure I understand this logic.

Don't you run the risk that the channel is empty, despite their being other tasks running which may fail at a later time?

Reply:
It is the Channel state at the time of the first Exception that counts.  The Channel being empty is a possibility, but the `while isready(c)` will just skip over it.  The Channel containing Tasks that fail at a later time is also a very real possibility - once we throw the first Exception(s) in the Composite, we stop looking at the rest.  But the only alternative is to wait for all Tasks to complete, which can cause the same hang we are trying to avoid in the first place.  (eg if one of the running Tasks is waiting for Channel input from a dead one.)  This PR doesn't get us true structured concurrency - it just guarantees _an_ Exception is thrown to help with debugging the system.